File: python/solutions/day_08.py
import math
from functools import reduce

from models.aoc_solution import AOCSolution, Dataset, Part

type Vec = tuple[int, int, int]


def vec_dist(a: Vec, b: Vec) -> float:
    x1, y1, z1 = a
    x2, y2, z2 = b
    return math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2 + (z1 - z2) ** 2)


class Day08(AOCSolution):
    EXPECTED = {
        Part.PART_ONE: {Dataset.SAMPLE: 40, Dataset.DATA: 32103},
        Part.PART_TWO: {Dataset.SAMPLE: 25272, Dataset.DATA: 8133642976},
    }

    def __post_init__(self) -> None:
        self.used: set[Vec] = set()
        self.vectors = self.parsed_data
        self.pairs = self.generate_pairs()
        self.circuits: list[set[Vec]] = []

    @property
    def parsed_data(self) -> list[Vec]:
        """Return coordinate tuples"""
        return [
            (int(x), int(y), int(z))
            for row in self.data.splitlines()
            for x, y, z in [row.split(",")]
        ]

    def generate_pairs(self) -> list[tuple[Vec, Vec]]:
        dists: dict[tuple[Vec, Vec], float] = {}
        for i, a in enumerate(self.vectors):
            for b in self.vectors[i + 1 :]:
                dist = vec_dist(a, b)
                dists[(a, b)] = dist
        return [key for key, _ in sorted(dists.items(), key=lambda pair: -pair[1])]

    def append_pair(self, a: Vec, b: Vec) -> None:
        """Append vectors to circuit & account for circuit joining"""
        for circ1 in self.circuits:
            if a in circ1:
                for circ2 in self.circuits:
                    if b in circ2:
                        if circ1 == circ2:
                            # Nothing happens
                            return
                        # Circuits join
                        self.circuits.remove(circ2)
                        circ1.update(circ2)
                        return
                # b was not in a circuit so it goes with a
                circ1.add(b)
                return
            if b in circ1:
                for circ2 in self.circuits:
                    if a in circ2:
                        if circ1 == circ2:
                            # Nothing happens
                            return
                        # Circuits join
                        self.circuits.remove(circ2)
                        circ1.update(circ2)
                        return
                # b was not in a circuit so it goes with a
                circ1.add(a)
                return
        # Neither were in a circuit so we get a new circuit
        self.circuits.append({a, b})
        return

    def part_one(self) -> int:
        """Largest 3 circuit sizes multiplied together"""
        counter = 10 if self.sample else 1000
        while counter:
            a, b = self.pairs.pop()
            self.append_pair(a, b)
            counter -= 1

        lengths = [len(circ) for circ in self.circuits]
        best3 = sorted(lengths, reverse=True)[:3]
        return reduce(lambda x, y: x * y, best3)

    def part_two(self) -> int:
        """Multiply x coords of final boxes to join all circuits"""
        while not self.circuits or len(self.circuits[0]) != len(self.vectors):
            a, b = self.pairs.pop()
            self.append_pair(a, b)
        return a[0] * b[0]


if __name__ == "__main__":
    Day08().run()

File: typescript/solutions/day_08.ts
import { isSample } from "../helpers/input";

type Vec = [number, number, number];

const parse = (input: string): Vec[] => {
  return input.split("\n").map((row) => row.split(",").map(Number) as Vec);
};

const dist = (a: Vec, b: Vec): number => {
  const [x1, y1, z1] = a;
  const [x2, y2, z2] = b;
  return Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2 + (z1 - z2) ** 2);
};

const generatePairs = (vectors: Vec[]): [Vec, Vec][] => {
  const combinations = vectors.flatMap((vec1, i) =>
    vectors.slice(i + 1).map((vec2) => [vec1, vec2] as [Vec, Vec]),
  );
  combinations.sort((a, b) => dist(...b) - dist(...a));
  return combinations;
};

class Playground {
  vectors: Vec[];
  circuits: Vec[][];
  pairs: [Vec, Vec][];

  constructor(input: string) {
    this.vectors = parse(input);
    this.circuits = [];
    this.pairs = generatePairs(this.vectors);
  }
  add_pair(): [Vec, Vec] {
    const [a, b] = this.pairs.pop()!;
    const matches = this.circuits.filter(
      (circ) => circ.includes(a) || circ.includes(b),
    );
    if (matches.length == 0) this.circuits.push([a, b]);
    else if (matches.length == 1) {
      if (!matches[0].includes(a)) matches[0].push(a);
      if (!matches[0].includes(b)) matches[0].push(b);
    } else {
      const [m1, m2] = matches;
      const i2 = this.circuits.indexOf(m2);
      this.circuits.splice(i2, 1);
      m1.push(...m2);
    }
    return [a, b];
  }
  get lengths(): number[] {
    const circuitLengths = this.circuits.map((circ) => circ.length);
    circuitLengths.sort((a, b) => b - a);
    return circuitLengths;
  }
  get isConnected(): boolean {
    return (
      !!this.circuits.length && this.circuits[0].length === this.vectors.length
    );
  }
}

export const part1 = (input: string): number => {
  const playground = new Playground(input);
  const count = isSample() ? 10 : 1000;
  for (let i = 0; i < count; i++) playground.add_pair();
  return playground.lengths.slice(0, 3).reduce((a, b) => a * b);
};

export const part2 = (input: string): number => {
  const playground = new Playground(input);
  let [a, b] = playground.add_pair();
  while (!playground.isConnected) [a, b] = playground.add_pair();
  return a[0] * b[0];
};
